src/dx/access/payloads.rs

```diff
@@ -0,0 +1,131 @@
+pub struct GrantTokenPayload<'request> {
+    /// How long (in minutes) the generated token should be valid.
+    pub ttl: usize,
+
+    /// A user ID, which is authorized to use the token to make API requests to
+    /// PubNub.
+    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
+    pub authorized_user_id: &'request Option<String>,
+
+    /// Extra metadata to be published with the request. Values must be scalar only.
+    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
+    pub meta: &'request Option<HashMap<String, MetaValue>>,
+
+    /// Permissions which should be granted to token.
+    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
+    pub permissions: Option<GrantTokenPermissionsPayload>,
+}
+
+impl<'request> GrantTokenPayload<'request> {
+    /// Create request payload.
+    ///
+    /// The information provided to the request builder is used to compose the
+    /// final payload, which should then be sent to the PubNub API endpoint.
+    pub(super) fn new<T, S, D>(request: &'request GrantTokenRequest<'_, T, S, D>) -> Self
+    where
+        T: Transport,
+        S: for<'se, 'rq> Serializer<'se, GrantTokenPayload<'rq>>,
+        D: for<'ds> Deserializer<'ds, GrantTokenResponseBody>,
+    {
+        GrantTokenPayload {
+            ttl: request.ttl,
+            authorized_user_id: &request.authorized_user_id,
+            meta: &request.meta,
+            permissions: Some(GrantTokenPermissionsPayload {
+                resources: resource_permissions(&request.resources),
+                patterns: resource_permissions(&request.patterns),
+            }),
+        }
+    }
+}
+
+/// Extract permissions for list of resources.
+fn resource_permissions(
+    resources: &Option<&[Box<dyn Permission>]>,
+) -> Option<GrantTokenResourcesPayload> {
+    if let Some(res) = resources {
+        let mut channels: HashMap<String, u8> = HashMap::new();
+        let mut groups: HashMap<String, u8> = HashMap::new();
+        let mut uuids: HashMap<String, u8> = HashMap::new();
+        res.iter().for_each(|perm| match perm.resource_type() {
+            ResourceType::Channel => {
+                channels.insert(perm.id().to_string(), *perm.value());
+            }
+            ResourceType::ChannelGroup => {
+                groups.insert(perm.id().to_string(), *perm.value());
+            }
+            ResourceType::UserId => {
+                uuids.insert(perm.id().to_string(), *perm.value());
+            }
+        });
+
+        return Some(GrantTokenResourcesPayload {
+            channels: if !channels.is_empty() {
+                Some(channels)
+            } else {
+                None
+            },
+            groups: if !groups.is_empty() {
+                Some(groups)
+            } else {
+                None
+            },
+            uuids: if !uuids.is_empty() { Some(uuids) } else { None },
+        });
+    }
+
+    None
```

Comment:
Proposal: You can use more "optional" syntax
```suggestion
    resources.map(|res| {
        let mut channels: HashMap<String, u8> = HashMap::new();
        let mut groups: HashMap<String, u8> = HashMap::new();
        let mut uuids: HashMap<String, u8> = HashMap::new();
        res.iter().for_each(|perm| match perm.resource_type() {
            ResourceType::Channel => {
                channels.insert(perm.id().to_string(), *perm.value());
            }
            ResourceType::ChannelGroup => {
                groups.insert(perm.id().to_string(), *perm.value());
            }
            ResourceType::UserId => {
                uuids.insert(perm.id().to_string(), *perm.value());
            }
        });

        GrantTokenResourcesPayload {
            channels: (!channels.is_empty()).then_some(channels),
            groups: (!groups.is_empty()).then_some(groups),
            uuids: (!uuids.is_empty()).then_some(uuids),
        }
    })
```